wiring/inc/spark_wiring_ble.h

```diff
@@ -0,0 +1,460 @@
+    MSB = 0,
+    LSB = 1
+};
+
+struct BleDeviceRoleType; // Tag type for BLE device role flags
+typedef Flags<BleDeviceRoleType, uint8_t> BleDeviceRoles;
+typedef BleDeviceRoles::FlagType BleDeviceRole;
+namespace ROLE {
+    const BleDeviceRole INVALID(BLE_ROLE_INVALID);
+    const BleDeviceRole PERIPHERAL(BLE_ROLE_PERIPHERAL);
+    const BleDeviceRole CENTRAL(BLE_ROLE_CENTRAL);
+    const BleDeviceRole ALL(0xff);
+}
+
+struct BleCharacteristicPropertyType; // Tag type for BLE attribute property flags
+typedef Flags<BleCharacteristicPropertyType, uint8_t> BleCharacteristicProperties;
+typedef BleCharacteristicProperties::FlagType BleCharacteristicProperty;
+namespace PROPERTY {
+    const BleCharacteristicProperty NONE(0);
+    const BleCharacteristicProperty READ(BLE_SIG_CHAR_PROP_READ);
+    const BleCharacteristicProperty WRITE(BLE_SIG_CHAR_PROP_WRITE);
+    const BleCharacteristicProperty WRITE_WO_RSP(BLE_SIG_CHAR_PROP_WRITE_WO_RESP);
+    const BleCharacteristicProperty NOTIFY(BLE_SIG_CHAR_PROP_NOTIFY);
+    const BleCharacteristicProperty INDICATE(BLE_SIG_CHAR_PROP_INDICATE);
+}
+
+namespace PARTICLE_BLE {
+    const uint8_t BLE_USER_DEFAULT_SVC_UUID[BLE_SIG_UUID_128BIT_LEN] = {
+        0x2d, 0x49, 0xc0, 0x26, 0xf0, 0xdb, 0xce, 0x06, 0x7a, 0x33, 0x6f, 0x12, 0x00, 0x00, 0xf3, 0x5d
+    };
+}
+
+typedef uint16_t BleConnectionHandle;
+typedef uint16_t BleAttributeHandle;
+
+typedef void (*OnDataReceivedCallback)(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context);
+typedef void (*OnScanResultCallback)(const BleScanResult* device);
+typedef void (*OnConnectedCallback)(const BlePeerDevice& peer);
+typedef void (*OnDisconnectedCallback)(const BlePeerDevice& peer);
+
+class BleAddress : public hal_ble_addr_t {
+public:
+    void operator = (hal_ble_addr_t addr) {
+        this->addr_type = addr.addr_type;
+        memcpy(this->addr, addr.addr, BLE_SIG_ADDR_LEN);
+    }
+
+    uint8_t operator [] (uint8_t i) const {
+        if (i >= BLE_SIG_ADDR_LEN) {
+            return 0;
+        }
+        return addr[i];
+    }
+
+    bool operator == (const BleAddress& addr) const {
+        if (this->addr_type == addr.addr_type && !memcmp(this->addr, addr.addr, BLE_SIG_ADDR_LEN)) {
+            return true;
+        }
+        return false;
+    }
+};
+
+class BleAdvertisingParams : public hal_ble_adv_params_t {
+};
+
+class BleConnectionParams : public hal_ble_conn_params_t {
+};
+
+class BleScanParams : public hal_ble_scan_params_t {
+};
+
+class BleCharacteristicHandles : public hal_ble_char_handles_t {
+public:
+    const BleCharacteristicHandles& operator = (const hal_ble_char_handles_t& halHandles) {
+        this->decl_handle = halHandles.decl_handle;
+        this->value_handle = halHandles.value_handle;
+        this->user_desc_handle = halHandles.user_desc_handle;
+        this->cccd_handle = halHandles.cccd_handle;
+        this->sccd_handle = halHandles.sccd_handle;
+        return *this;
+    }
+};
+
```

Comment:
Let's just in case add static assertion for all this classes that inherit from hal structs and may be passed to ble hal directly.
```cpp
static_assert(std::is_pod<BleAdvertisingParams>::value, "BleAdvertisingParams is not a POD struct");
```